### ISSUE TITLE
OpenCV CMake windows build problem with PYTHON3_LIBRARY 

### ISSUE THREAD

**Issue by quickgrid** [opencv/opencv#20206]
##### System information (version)

- OpenCV => 4.5.2
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2019

##### Detailed description

I am trying compile opencv for python 3.9 in conda. Initially got `no module named cv2` error after copying `cv2.cp39-win_amd64.pyd` file to `Lib/site-packages`. Then tried visual studio cmake install target that gave `ImportError: DLL load failed while importing cv2: The specified module could not be found.` error.

I have unticked `BUILD_SHARED_LIBS`, ticked `BUILD_opencv_world` and set `CMAKE_CONFIGURATION_TYPES` to `RELEASE` only.  Also set path for opencv-contrib via `OPENCV_EXTRA_MODULES_PATH` to opencv-contrib source modules folder. Ticked `WITH_CUDA`, `OPENCV_DNN_CUDA`, set `CUDA_ARCH_BIN` value based for my GPU. If the cuda part is not used then copying the pyd file to site-packages worked. In this case, I modified `CMakeCache.txt` to replace all python 3.7 path to correct conda env path.

![1](https://user-images.githubusercontent.com/1857293/120619558-e660ed00-c47d-11eb-84d5-e51fe6d9acda.png)

Now I am noticing no matter what I set in `PYTHON3_LIBRARY` it is not showing up in configure output. I have tried CMake `3.18` and `3.20.3`. I clicked configure multiple times after changing the path but remains same. After clicking configure first time, `PYTHON3_LIBRARY` is set to `C:/Users/computer/miniconda3/libs/python37.lib` and does not change anymore. Pointing the path to another env `python37.lib` does not help. Deleting cmake cache and reconfiguring, using another folder does not help.

This is `cv2` folder in under `Lib/site-packages/` does not have config for python 3.9 after install. Doing an install with Visual Studio `CMakeTargets` installs like this, 

![3](https://user-images.githubusercontent.com/1857293/120618901-528f2100-c47d-11eb-9e55-ce48dbb3106d.png)

![2](https://user-images.githubusercontent.com/1857293/120618928-591d9880-c47d-11eb-8fa5-fc4e604fc564.png)

 

##### Steps to reproduce

Have anaconda/miniconda with multiple environments and python 3.9 with numpy. Try to set `PYTHON3_LIBRARY` to different path in cmake configure.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by alalek** [comment]
Check this thread: #19972

> CUDA

You need to add CUDA's DLLs paths into `os.add_dll_directory()` before importing cv2.

**Comment by quickgrid** [comment]
Thank you @alalek  for the suggestion. I tried `os.add_dll_directory()` it did not work. Still showing, `ImportError: DLL load failed while importing cv2: The specified module could not be found.`.

```
import os
os.add_dll_directory(r"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0")
os.add_dll_directory(r"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\bin")
os.add_dll_directory(r"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\libnvvp")
os.add_dll_directory(r"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\extras\CUPTI\lib64")
os.add_dll_directory(r"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include")
os.add_dll_directory(r"C:\portable\opencv-compilation\build\install\x64\vc16\bin")
os.add_dll_directory(r"C:\portable\opencv-compilation\build\bin\Release")
os.add_dll_directory(r"C:\portable\opencv-compilation\build\3rdparty\ffmpeg")
```

Following, #19972 using [dependencies](https://github.com/lucasg/Dependencies) utility I see it is looking for `python37.dll` even though in cmake I stated python 3.9 for all and ran configure. I have hovered over all `*.dll` and it says `python37.dll` not found in disk. All others are loaded correctly.

![1](https://user-images.githubusercontent.com/1857293/120681651-69556800-c4bd-11eb-9245-4f0533c6b3ca.png)


This is the same problem I mentioned above. I have several conda environments setup with base one being python 3.7. When I run configure first time in CMake GUI it finds the default python 3.7. Then even if change flag values of `PYTHON3_LIBRARY` in many place of `CMakeCache.txt` it was referencing my the base python 3.7 version instead of python 3.9. Also in CMake output window it always shows wrong path in the `Libraries` part seen in my first image.

I will try one more time with manually correcting all paths in `CMakeCache.txt` and running build.

**Comment by quickgrid** [comment]
Manually editing `CMakeCache.txt` to fix all correct path references worked. I am also getting code completion in pycharm. I did not need to include any other dll or `os.add_dll_directory()`. These are the steps I tried,

```
0 : UNTICK
1 : TICK
... :  Path in pics above

Install zlib for python 3.9 in addition to numpy via conda-forge.

<configure>

FAST_MATH = 1
BUILD_SHARED_LIBS = 0
BUILD_opencv_world = 1
OPENCV_DNN_CUDA = 1
WITH_CUDA = 1
BUILD_JAVA = 0
BUILD_opencv_java_bindings_generator = 0
OPENCV_EXTRA_MODULES_PATH = ...
PYTHON3_EXECUTABLE = ...
PYTHON3_INCLUDE_DIR = ...
PYTHON3_LIBRARY = ...
PYTHON3_NUMPY_INCLUDE_DIRS = ...
PYTHON3_PACKAGES_PATH = ...

<configure>

CUDA_ARCH_BIN = [number based on gpu]
CUDA_FAST_MATH = 1
WITH_TESSERECT = 0

<configure>

Close cmake. 
Manually fixed CMakeCache.txt to correct opencv path, version, numpy path, version. 
All conda base env path replaced with custom env path.
Reopen cmake.

<configure>

<generate>
```

After compilation I copied the pyd file to site-packages and it worked without error on multiple different codes. Test code,

```
import cv2
print(cv2.__version__)
count = cv2.cuda.getCudaEnabledDeviceCount()
print(count)
```

**Output**
```
4.5.2
1
```

I see in the `build > python_loader > CV2` folder there is `config-3.7` but none for `3.9` so I rather copied pyd to site-packages than running INSTALL through visual studio. Editing the file contents and folder names also work if running install.

I am wondering what is the problem here. What is the correct way of easily compiling opencv with cuda without going into so much manual editing?

**Comment by alalek** [comment]
> there is config-3.7 but none for 3.9

CMake process should be checked.
Ensure that you clear CMake cache (or build directory) before changing of build options / dependencies (cache caches old values and don't check again).

In general, prefer to use "cmake" from the command line (as single command) instead of `cmake-gui` which has design issues (because of cache).

**Comment by quickgrid** [comment]
Thank you I will go with command line from now on. Seems easier than setting default python environment variables or cleaning cache each time for cmake-gui. 

On this note, I did try deleting whole build folder after running configure on gui once, then changing flags then rerunning configure. But it still found conda base python37.lib. I did not have python environment variable setup. Deleting `CMakeFiles` only within build folder is also same. I also tried cmake `File > Delete Cache` but end result was same. Maybe I am doing it wrong.

As I was able to compile successfully with manual editing and using cmake from cmd is a better option I am closing this.
